### DISCUSSION TITLE
Flatten semi-transparent areas and side effects

### DISCUSSION THREAD

**Discussion by loicaigon** [ImageMagick/ImageMagick#6122] [General] [answered]
Hi there,

I did a search on this topic but found no entries that would help so here I am. But if I missed something, let me know.

Here is my issue, say we have a round with a shadow as PNG and convert it to GIF. I have succeeded in "flattening" the shadow with a background color AND having the GIF transparent with this command:
`convert \( -background "#FF0000" -flatten "green.png" \) -transparent "#FF0000" "redOut.gif"`
so 
<img width="300" alt="green" src="https://user-images.githubusercontent.com/861261/222207687-8dc51aa9-94a7-4560-bdfd-49d8bc90d3d1.png">
exits fine as
![redOut](https://user-images.githubusercontent.com/861261/222207805-fda6a0e1-18ad-42a7-b280-095fe0efddc6.gif)

My problem comes when the main graphic can have the color included in the design for ex:
<img width="300" alt="red" src="https://user-images.githubusercontent.com/861261/222208022-ce92880f-5e02-4f7c-9805-4d3a7c7037f7.png">
which becomes
![redOut](https://user-images.githubusercontent.com/861261/222208216-faddb471-2be1-414f-9864-8331085fd82a.gif)

I can understand it's not a bug as conceptually IM is doing what is asked. But what I need to achieve would be something more like: 

![Capture d’écran 2023-03-01 à 17 55 59](https://user-images.githubusercontent.com/861261/222208514-27834492-70f6-4016-96fe-5cc7c4687279.png)

Where I can preserve the color in the design and yet flatten transparent areas of the PNG.

Is that achievable?

Thanks in advance for the help.

**Comment by fmw42** [comment]
This is not clear to me what you want.  Your command:

> convert \( -background "#FF0000" -flatten "green.png" \) -transparent "#FF0000" "redOut.gif"

Is not a valid command. Where is the input?

Please post your exact command and your input and output images so that we can test with it.  Also say exactly what you want to do or post your expected result.

**Comment by loicaigon** [comment]
Hi,

Sorry if this was unclear. Basically what I want as an output is a "transparent" GIF where any transparent pixels in the PNG source file get flattened with a specific color. So 100% transparent in the PNG is 100% Transparent in the GIF, any other value is a mix between the transparency value of the pixel and a specific color.

I am attaching a zip which contains:
- 3 png I want to animate as a GIF
- a file named "ideal_output.gif" which shows the kind of result I want. This one works because the light blue backgound is not to be found in the PNG files
- a file names "red_problem.gif" which shows the issue when the color defined for the transparency in the GIF can be found in the PNG files. Here, the red pixels in the red png simply disappear.

I wish I can use a color for flattening transparent area without affecting 100% opaque pixels with same value. I hope this will make things clearer.

For the command, I tried to simplify it, I put it the full one here:
`convert \( -delay 25 -background "#FF0000" -flatten "red.png" \) -set dispose 2 \( -delay 25 -background "#FF0000" -flatten "green.png" \) -set dispose 2 \( -delay 25 -background "#FF0000" -flatten "blue.png" \) -set dispose 2 -loop 0 -transparent "#FF0000" "output.gif"`




[Archive.zip](https://github.com/ImageMagick/ImageMagick/files/10863709/Archive.zip)

**Comment by fmw42** [comment, marked as answer]
The solution is not to make your background exactly the same as the color of the image. For example note the background color is slightly different from pure red; that is use #FF0011 for example rather than #FF0000. This will not be seen since your then make that color transparent:

`convert \( "red.png" -background "#FF0011" -flatten +write t2.png \) -transparent "#FF0011" "result_red.gif"`

![result_red](https://user-images.githubusercontent.com/13458218/222258333-4cc51ad6-87cf-4ca7-8699-8e5d3cf9f937.gif)

Note that proper IM syntax, reads the input before doing operations such as -flatten. But IM 6 is forgiving, so it works in your order. But IM 7 would not work well as it is more strict in the use of proper syntax.

**Reply by loicaigon** [reply]
That seems like a good approach indeed. At first, I was a bit concerned about color mismatches (as when the GIF would be placed on a red-colored HTML page) but when I tried this, I couldn't notice a difference. Thanks a lot for the help today.

**Reply by fmw42** [reply]
Should work with #FF0001. Give it a try and see if that works also.

**Reply by loicaigon** [reply]
Yes, that definitely works. I just shift by a bit and that's perfect.
